### PR TITLE
Add subtyping feature and improvements to the Ballerina bindgen tool

### LINKED ISSUE #22089: Support for Java subtyping in the Ballerina bindgen tool
**Description:**
Currently, the bindgen tool does not support Java subtyping. This needs to be implemented taking into account Java Interfaces, Java abstract classes, and parent classes.

For instance, the current approach of assigning a `java.io.FileInputStream` object into an `java.io.InputStream` object is as follows.
```ballerina
FileInputStream fileInput = check newFileInputStream2(filename);
InputStream inputStream = new(fileInput.jObj);
```
However, the tool should be able to identify `java.io.FileInputStream` as a subytpe of `java.io.InputStream` and allow the below assignment.
```ballerina
InputStream inputStream = check newFileInputStream2(filename);
```

**Steps to reproduce:**

**Affected Versions:**
Ballerina 1.2.0

**OS, DB, other environment details and versions:**

**Related Issues (optional):**

### PR DESCRIPTION
**Pull request by IrushiL**
## Purpose
This PR consists of the following improvements and issue fixes to the Ballerina bindgen tool.

**Improvements:**
* Provide support for Java subtyping.
* An improvement to the folder structure of the generated bindings.
* Improvements to error messages and information provided to the user.

**Fixes:**
* Fixes an issue in the generation of child classes.
* Fixes an issue that occurs while reading constants file when re-running the tool.
* Fixes some issues related to type mappings.

Fixes #22089 

## Check List 
- [X] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: misc/ballerina-bindgen/src/main/java/org/ballerinalang/bindgen/command/BindingsGenerator.java [hunk 1]
```diff
@@ -56,90 +58,121 @@
...
+            }
+        } catch (BindgenException e) {
+            throw new BindgenException("Error while loading the classpaths.", e);
+        }
+        return classLoader;
+    }
+
+    private void setDirectoryPaths() {
+
+        if (outputPath == null) {
```
Comment (warunalakshitha) [inline misc/ballerina-bindgen/src/main/java/org/ballerinalang/bindgen/command/BindingsGenerator.java:130]: If you use temp variable to hold the outputPath var and set its values to userDir if null, then we get remove duplicate codes.

File: misc/ballerina-bindgen/src/main/java/org/ballerinalang/bindgen/model/JParameter.java [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+        }
+        externalType = getBallerinaHandleType(parameterClass);
+        fieldName = "arg";
+    }
+
+    JParameter(Parameter parameter) {
+
+        this(parameter.getType());
+        fieldName = parameter.getName();
+
```
Comment (warunalakshitha) [inline misc/ballerina-bindgen/src/main/java/org/ballerinalang/bindgen/model/JParameter.java:81]: extra new line. You also can remove the extra new lines after each method signature as well.

File: misc/ballerina-bindgen/src/main/java/org/ballerinalang/bindgen/command/BindingsGenerator.java (hunk 1 above)
Comment (IrushiL) [reply misc/ballerina-bindgen/src/main/java/org/ballerinalang/bindgen/command/BindingsGenerator.java]: Simplified the code with a temp variable.

File: misc/ballerina-bindgen/src/main/java/org/ballerinalang/bindgen/model/JParameter.java (hunk 2 above)
Comment (IrushiL) [reply misc/ballerina-bindgen/src/main/java/org/ballerinalang/bindgen/model/JParameter.java]: Done.

**Comment by warunalakshitha** [comment]
IMO it is better add some tests to cover some use cases of bindgen tool.

**Pull request merged**